src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/RoundaboutMissingTagCheck.java

```diff
@@ -0,0 +1,193 @@
+    // Minimum intersection with Navigable Roads
+    private static final int MINIMUM_INTERSECTION = 2;
+    private static final int MODULUS = 10;
+    private static final int FIRST_EDGE_SECTION = 1;
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(MISSING_JUNCTION_TAG_INSTRUCTION);
+    private static final double MAX_THRESHOLD_DEGREES_DEFAULT = 40.0;
+    private static final double MIN_THRESHOLD_DEGREES_DEFAULT = 10.0;
+    private final Angle maxAngleThreshold;
+    private final Angle minAngleThreshold;
+
+    /**
+     * Default constructor
+     *
+     * @param configuration
+     *            the JSON configuration for this check
+     */
+    public RoundaboutMissingTagCheck(final Configuration configuration)
+    {
+        super(configuration);
+        this.maxAngleThreshold = configurationValue(configuration, "threshold.degrees",
+                MAX_THRESHOLD_DEGREES_DEFAULT, Angle::degrees);
+        this.minAngleThreshold = configurationValue(configuration, "threshold.degrees",
+                MIN_THRESHOLD_DEGREES_DEFAULT, Angle::degrees);
+    }
+
+    /**
+     * Validates if given {@link AtlasObject} is actually an {@link Edge} and is a potential
+     * roundabout and also corresponding OSM identifier shouldn't be flagged before (this is for
+     * avoiding duplicate flags)
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge && !this.isFlagged(object.getOsmIdentifier())
+                && object.getIdentifier() % MODULUS == FIRST_EDGE_SECTION
+                && ((Edge) object).isMasterEdge() && HighwayTag.isCarNavigableHighway(object)
+                && object.getTag(JunctionTag.KEY).isEmpty() && object.getTag(AreaTag.KEY).isEmpty()
+                && isPartOfClosedWay((Edge) object) && intersectingWithMoreThan((Edge) object);
+    }
+
+    /**
+     * Flags an {@link Edge} is its circular shape and connected to at least
+     * {@link RoundaboutMissingTagCheck#MINIMUM_INTERSECTION} navigable roads. {@link Edge} doesn't
+     * have junction=roundabout or area=yes tags.
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+
+        final PolyLine originalGeom = buildOriginalOsmWayGeometry(edge);
+        // check maximum angle
+        final List<Tuple<Angle, Location>> maxOffendingAngles = originalGeom
+                .anglesGreaterThanOrEqualTo(this.maxAngleThreshold);
+        // check minimum angle
+        final List<Tuple<Angle, Location>> minOffendingAngles = originalGeom
+                .anglesLessThanOrEqualTo(this.minAngleThreshold);
+
+        if (maxOffendingAngles.isEmpty() && minOffendingAngles.isEmpty())
+        {
+            this.markAsFlagged(object.getOsmIdentifier());
+            return Optional.of(createFlag(new OsmWayWalker(edge).collectEdges(),
```

Comment:
We could cache the waywalker call at the beginning of this method

Reply:
Or potentially in the validCheckForObject method, as I see now the usage is out of this scope

Reply:
yes. tested it, however it significantly slowing down the performance. 
compilation time. tested against China data on local:
1 - use WayWalker in each method - 7,48 min (current)
2 - cache WayWalker and pass Set Edges to every method - 9,64 min